raft/node.go

```diff
@@ -118,6 +118,10 @@ type Node interface {
 	Campaign(ctx context.Context) error
 	// Propose proposes that data be appended to the log.
 	Propose(ctx context.Context, data []byte) error
+	// ProposeWithCancel proposes that data be appended to the log
+	// and will invoke the cancel() if the proposal is dropped.
+	// Application can use this to fail fast when the proposal is dropped or cancelled.
+	ProposeWithCancel(ctx context.Context, cancel context.CancelFunc, data []byte) error
```

Comment:
why do we need to add this? can we improve Propose func so that it would return an error if proposal is dropped?

Reply:
Currently, the `Propose` just send the proposal to chan, and will not wait. If we change the behavior for `Propose` to block and return an error it may affect others.

I think that it may be better to add a new method. Any suggestion?

Reply:
Mostly, cancel function is used to cancels the child outside, like:

```go
ctx, cancel := WithCancel(parent)
pass(ctx)
cancel()
```
So it is strange for me to see this style here, maybe a callback is better.

Reply:
Yeah,  the method name needs to be improved. A new propose method name should make the caller known that a cancel callback will be used to get notified when the proposal is canceled by dropping. (I use the function `context.CancelFunc` as callback since It is actually canceled the `context.Context`)

Reply:
> Currently, the Propose just send the proposal to chan, and will not wait.

that is true. however, step func should be fast and non-blocking most of the time. changing Proposal to wait a bit longer for stepping to finish should have very a little impact if any.

Maybe I missed something. What is your concern specifically?

Reply:
@siddontang 

exactly. if we want to make it work, it will return either a callback or a chan, which will be closes if canceled.